forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt

```diff
@@ -0,0 +1,187 @@
+
+    @JvmOverloads constructor(): super()
+
+    /**
+     * The optional userId
+     */
+    private var userId: String? = null
+
+    /**
+     * The challenge received from server
+     */
+    private lateinit var challenge: String
+    /**
+     * The title to be displayed in biometric prompt
+     */
+    private lateinit var title: String
+    /**
+     * The subtitle to be displayed in biometric prompt
+     */
+    private lateinit var subtitle: String
+    /**
+     * The description to be displayed in biometric prompt
+     */
+    private lateinit var description: String
+    /**
+     * The timeout to be to expire the biometric authentication
+     */
+    private var timeout: Int? = null
+
+    private val tag = DeviceBindingCallback::class.java.simpleName
+
+    override fun setAttribute(name: String, value: Any) = when (name) {
+        "userId" ->  userId = value as? String
+        "challenge" -> challenge = value as String
+        "title" -> title = value as? String ?: ""
+        "subtitle" -> subtitle = value as? String ?: ""
+        "description" -> description = value as? String ?: ""
+        "timeout" -> timeout = value as? Int
+        else -> {}
+    }
+
+    override fun getType(): String {
+        return "DeviceSigningVerifierCallback"
+    }
+
+    /**
+     * Input the JWS key to the server
+     * @param value The JWS value.
+     */
+    fun setJws(value: String?) {
+        super.setValue(value, 0)
+    }
+
+
+    /**
+     * Input the Client Error to the server
+     * @param value DeviceBind ErrorType .
+     */
+    fun setClientError(value: String?) {
+        super.setValue(value, 1)
+    }
+
+    /**
+     * Bind the device.
+     *
+     * @param context  The Application Context
+     * @param listener The Listener to listen for the result
+     */
+    @JvmOverloads
+    fun sign(context: Context,
+             listener: FRListener<Void>) {
+        execute(context, listener = listener)
+    }
+
+    /**
+     * Helper method to execute binding , signing, show biometric prompt.
+     *
+     * @param context  The Application Context
+     * @param listener The Listener to listen for the result
+     * @param authInterface Interface to find the Authentication Type
+     * @param encryptedPreference Persist the values in encrypted shared preference
+     */
+    protected open fun authenticate(userKey: UserKey,
+                             keyAware: KeyAware = KeyAware(userKey.userId),
+                             listener: FRListener<Void>) {
+        val authInterface =  getDeviceBindInterface(userKey, keyAware)
+        if(authInterface.isSupported().not()) {
+            handleException(Unsupported(), listener)
+            return
+        }
+        try {
+            authInterface.authenticate(timeout ?: 60) { result ->
+                if (result is Success) {
+                    val privateKey = keyAware.getPrivateKey(userKey.keyAlias)
```

Comment:
Do we need the keyAware to retrieve the private key? If the authenticator get a hold of the keyAaware object, we may able to simplify the logic here by not passing the private key around.

Reply:
Good point , we can wrap this in the authenticator